library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java

```diff
@@ -1416,6 +1450,32 @@ private Metadata getAdjustedMetadata(@Nullable Metadata metadata) {
       }
       return new Metadata(newMetadataEntries);
     }
+
+    @Override
+    public void sampleMetadata(
+        long timeUs,
+        @C.BufferFlags int flags,
+        int size,
+        int offset,
+        @Nullable CryptoData cryptoData) throws IOException {
+      // TODO - chunkless prepare, sampleQueue list is not yet initialized for first chunk
+  //    Assertions.checkNotNull(chunk, "sampleMetadata without a MediaChunk?");
+      if (chunk == null) {
```

Comment:
I don't think I understand when this can happen, and the TODO above didn't help me figure it out. Could you elaborate?

Reply:
The first segments to load when it is not [chunk-less prepare mode](https://medium.com/google-exoplayer/faster-hls-preparation-f6611aa15ea6).  This is because until the prepare is complete the `SampleStream` is not added to the `HlsSampleStreamWrapper.sampleQueues[]` array and the `HlsSampleStreamWrapper.init()` method does not get called until the load has already started.  So samples can be committed before the current loading `MediaChunk` is set.  

I'm working out how architecturally to make this happen, so the TODO is for me.

Reply:
Sorry, I still don't understand this. It looks to me like `HlsSampleStreamWrapper.init()` is always called before a chunk starts to load. Can you actually reproduce this not happening, and if so, how?

It's also unclear whether this is about chunk-less or non-chunk-less preparation. The comment says one thing, but the conversation says another. Please clarify; thanks!

Reply:
I see how this happens now.

Reply:
I fixed this as part of the merge.